### ISSUE TITLE
Error in /device-dynamic resource

### ISSUE THREAD

**Issue by rosogon** [mF2C/cimi#44]
A `GET /device-dynamic` does not return the count field (this is happening on Leader, where I am creating the device-dynamics). I am POSTing a new device-dynamic and cimi responds with 201. The resource is there if I directly GET by Id. But it is not shown on a `GET /device-dynamic`.

```{
  "acl" : {
    "owner" : {
      "principal" : "ADMIN",
      "type" : "ROLE"
    },
    "rules" : [ {
      "principal" : "USER",
      "type" : "ROLE",
      "right" : "MODIFY"
    } ]
  },
  "resourceURI" : "http://schemas.dmtf.org/cimi/2/DeviceDynamicCollection",
  "id" : "device-dynamic",
  "operations" : [ {
    "rel" : "add",
    "href" : "device-dynamic"
  } ],
  "deviceDynamics" : [ ]
}
```

The same `GET /device-dynamic` on child returns the same as in leader, but adding the count field:

```
{
  "count" : 0,
  "acl" : {
    "owner" : {
      "principal" : "ADMIN",
      "type" : "ROLE"
    },
    "rules" : [ {
      "principal" : "USER",
      "type" : "ROLE",
      "right" : "MODIFY"
    } ]
  },
  "resourceURI" : "http://schemas.dmtf.org/cimi/2/DeviceDynamicCollection",
  "id" : "device-dynamic",
  "operations" : [ {
    "rel" : "add",
    "href" : "device-dynamic"
  } ],
  "deviceDynamics" : [ ]
}
```

**Comment by rosogon** [comment]
Anna: for all the functions that CIMI sends to dataClay (create, read, update, delete), we are receiving the type in the "CIMI format" (i.e. "device", "device-dynamic", etc). However, for the query the type we receive is in the format "Device", "DeviceDynamic", etc, while we are expecting the same as for the rest of functions. This still works (by chance) in most cases, that's why some queries work (and that's why it was hard to find). But for those resources that have links to other resources, the query fails

**Comment by rosogon** [comment]
This bug is expected to be solved when #42 is finished.

**Comment by rosogon** [comment]
needs #42

**Comment by cjdcordeiro** [comment]
if this is the one you were mentioning @rosogon , then no. I'm not sure this is related with the integration as there is no actual error being thrown

**Comment by aqueralt** [comment]
Well, this behaviour happened because the value of the "type" parameter we get from the proxy in the query method (and only in this case) is not CIMI-like ("device-dynamic") as in the rest of methods, but java-like ("DeviceDynamic"). Is it still like this in the new proxy? And if so, would it be possible to get the types as in the rest of methods ("device-dynamic", etc)?

**Comment by cjdcordeiro** [comment]
@aqueralt that's not it. All queries follow a PascalCase convention. So y it is true that you see DeviceDynamic, but you also see ServiceInstace, and so on...and it works for others.

The real problem here is that device-dynamic uses resource links, which means there is at least one attribute that points to another existing one (in this case, `device` and `myLeaderID`. 

If you look at the logs from dcproxy, you'll see we send the right query, but then in return, for some reason, DataClay is replacing the values of all resource links by the full internal data model, and not the actual value of that attribute. Basically if you reproduce this and you look at the dcproxy logs, you'll see that dcproxy has received 11k lines of JSON from DataClay.

**Comment by aqueralt** [comment]
@cjdcordeiro Yes, I know, that's what I said at some point, and @rosogon pasted in a comment above. I mean that for all the methods (create, read, update, delete) we are getting the types as they appear in the CIMI resource specification ("device-dynamic" or whatever), but not in the query, and we are using the same function everywhere to recognize subobjects and return the resource link instead of the whole data of the linked object in dataClay. 
I thought this difference of type format in CRUD vs query was not intended, and that in the new integration all the types we get in all methods would look the same. Since this is not the case, we can adapt to the specific case of the query, or the proxy can give us the type in the same format as it does for the rest of methods. Just choose what you think is the best option.

**Comment by cjdcordeiro** [comment]
sry I don't understand. It's a simple read (GET). I see this in the dcproxy logs:

`INFO: dispatching dataclay function:  :add ({:serviceInstance`
`INFO: dispatching dataclay function:  :query (ServiceInstance`

and it works, for every resource which doesn't have a `resource-link`, so what does it have to do with the type of action or type of resource?

**Comment by aqueralt** [comment]
In dataClay we don't store JSONs but we store instances of Java objects corresponding to each type of resource, linked by java references in case they have links to other resources. For the integration with CIMI, we built a wrapper on top of dataClay that transforms JSON<->Java/dataClay so that you could call us as similar as possible to what you did with ElasticSearch. 
Then, regarding the issue: in the "read" method, for instance, we get the object from dataClay and replace the references (if any) by the corresponding resource link. The function we use to do this needs to be aware of the types of the objects, and it assumes they are like "device-dynamic". 
For the query, we use the same function to build the response because we understood that types would always look the same. As you have seen, the error is only when there are subobjects, since this is the only case when we need to do some post-processing before building the JSON (when there are no subobjects, we just build a JSON directly from what we get from dataClay). 
So this is the thing, hope I explained it better now. As I said, we can adapt our code to deal with queries differently, or if you just give us the types in "CIMI-like format", it will just work.

**Comment by cjdcordeiro** [comment]
fixed with https://github.com/mF2C/dataClay/pull/34#event-2111556259

@rosogon try dcproxy:2.5

**Comment by rosogon** [comment]
Hi,

This exact case (creating a device-dynamic on a leader) is fixed with dcproxy:2.5. 

Thanks @aqueralt and @cjdcordeiro.

Closing.
